Add blinded contact records to Contacts and ConvoInfoVolatile

mpretty-cyro opened the pull request:
Adds blinded contact records to `Contacts` and `ConvoInfoVolatile` to simplify syncing them between devices and allow `libSession` to construct the full conversation list

~~**Note:** This is based on #51~~

mpretty-cyro committed:
Added blinded contact records to Contacts and ConvoInfoVolatile

jagerman commented on include/session/config/contacts.h:
```diff
@@ -36,6 +36,31 @@ typedef struct contacts_contact {
+                          // only has port if non-default, has trailing / removed)
+    unsigned char pubkey[32];  // 32 bytes (not terminated, can contain nulls)
+
+    char name[101];  // This will be a 0-length strings when unset
```
```suggestion
    char name[101];  // This will be a 0-length string when unset
```

jagerman commented on include/session/config/contacts.hpp:
```diff
@@ -44,8 +46,24 @@ namespace session::config {
 ///     j - Unix timestamp (seconds) when the contact was created ("j" to match user_groups
 ///         equivalent "j"oined field). Omitted if 0.
+///
+/// b - dict of blinded contacts.  This is a nested dict where the outkey keys are the BASE_URL of
```
```suggestion
/// b - dict of blinded contacts.  This is a nested dict where the outer keys are the BASE_URL of
```

jagerman commented on include/session/config/contacts.hpp:
```diff
@@ -44,8 +46,24 @@ namespace session::config {
+///     `R` - dict of blinded contacts from the server; each key is the blinded session pubkey
+///     without the prefix ("R" to match user_groups equivalent "R"oom field, and to make use of
+///     existing community iterators, binary, 32 bytes), value is a dict containing keys:
+///       containing keys:
```
```suggestion
```

jagerman commented on include/session/config/contacts.hpp:
```diff
@@ -339,6 +398,70 @@ class Contacts : public ConfigBase {
     bool accepts_protobuf() const override { return true; }
 
+  protected:
+    // Drills into the nested dicts to access open group details
```
```suggestion
    // Drills into the nested dicts to access community details
```

jagerman commented on include/session/config/contacts.hpp:
```diff
@@ -97,6 +115,47 @@ struct contact_info {
 };
 
+struct blinded_contact_info : community {
+    using community::community;
```
Inheriting from community here feels a little strange, because a blinded contact does not satisfy "is a community" but rather "has a community", and so I think you could do everything here more cleanly by just having a `community comm;` member instead of inheritance.

mpretty-cyro replied:
So the main reason for the inheritance was so I could reuse the `comm_iterator_helper` - I've updated the struct to have a `community comm;` instead and added the `set_{x}` functions to redirect to setting the values on the member (not sure if this will work as desired or not 😬)

mpretty-cyro committed:
Comment fixes from PR feedback

mpretty-cyro committed:
Updated `blinded_contact_info` to not inherit from `community`

mpretty-cyro committed:
Added 'profile_seqno' to contact_info & member to version profile info

mpretty-cyro committed:
Changed 'profile_seqno' to 'profile_updated'

mpretty-cyro committed:
Ran the formatter, fixed a network test that was still talking to testnet

mpretty-cyro committed:
Ran the formatter again

mpretty-cyro committed:
Added unit tests for blinded contacts, bug fixes and cleanup

mpretty-cyro committed:
Fixed a test failure caused by odd syntax

jagerman committed:
Make profile timestamp type safe std::chrono::sys_seconds

This uses std::chrono::sys_seconds for the new profile_updated timestamp
field, so that we can avoid the unfortunate bug that happened in earlier
releases where some clients violated the documented restrictions and
passed milliseconds or microseconds: sys_seconds is absolutely
unambiguous as to what it holds and makes such an error virtually
impossible.

- Adds `session::to_sys_seconds(int)` that can get a proper sys_seconds
  from a maybe-s, ms, or µs value by guessing based on the magnitude.

- Add methods for getting/setting sys_seconds timestamps from/into
  configs.

jagerman committed:
Add `int_or_0`/`string_or_empty` to match ts_or_epoch

There are a lot of internal calls such as `maybe_int(c,
"key").value_or(0)` and `maybe_string(c, "key").value_or("")`.  This
makes it them slightly less cumbersome by adding `int_or_0` and
`string_or_empty` functions that handle the fallback-to-0/empty
automatically.

jagerman committed:
Drone: drop unneeded backports from Debian 12 build

jagerman committed:
blinded_contact_info: type-safe timestamp

jagerman commented:
Rebased on top of #54 with an update applied to use the same type-safe timestamp.

mpretty-cyro committed:
Bumped version number
